app/models/product_importer.rb

```diff
@@ -0,0 +1,314 @@
+require 'roo'
+
+class ProductImporter
+  extend ActiveModel::Naming
+  include ActiveModel::Conversion
+  include ActiveModel::Validations
+
+  def initialize(file, editable_enterprises, options={})
+    if file.is_a?(File)
+      @file = file
+      @options = options
+      @sheet = open_spreadsheet
+      @valid_entries = {}
+      @invalid_entries = {}
+
+      @products_to_create = []
+      @variants_to_create = []
+      @variants_to_update = []
+      
+      @products_created = 0
+      @variants_created = 0
+      @variants_updated = 0
+
+      @editable_enterprises = {}
+      editable_enterprises.map { |e| @editable_enterprises[e.name] = e.id }
+
+      validate_all
+    else
+      self.errors.add(:importer, "error: no file uploaded")
+    end
+  end
+
+  def persisted?
+    false #ActiveModel, not ActiveRecord
+  end
+
+  # Private methods below which only work with a valid spreadsheet object can be called publicly
+  # via here if the spreadsheet was successfully loaded, otherwise they return nil (without error).
+  def method_missing(method, *args, &block)
+    if self.respond_to?(method, include_private=true)
+      @sheet ? self.send(method, *args, &block) : nil
+    else
+      super
+    end
+  end
+
+  private
+
+  def open_spreadsheet
```

Comment:
Hm, interesting approach. I can't see anything wrong with it, looks like you've covered off all of the bases.... 👍